base/abstractarray.jl

```diff
@@ -287,6 +287,138 @@ size_to_strides(s, d) = (s,)
 size_to_strides(s) = ()
 
 
+abstract type MemoryLayout end
+struct UnknownLayout <: MemoryLayout end
+abstract type AbstractStridedLayout <: MemoryLayout end
+abstract type AbstractIncreasingStrides <: AbstractStridedLayout end
+abstract type AbstractColumnMajor <: AbstractIncreasingStrides end
+struct DenseColumnMajor <: AbstractColumnMajor end
+struct ColumnMajor <: AbstractColumnMajor end
+struct IncreasingStrides <: AbstractIncreasingStrides end
+abstract type AbstractDecreasingStrides <: AbstractStridedLayout end
+abstract type AbstractRowMajor <: AbstractDecreasingStrides end
+struct DenseRowMajor <: AbstractRowMajor end
+struct RowMajor <: AbstractRowMajor end
+struct DecreasingStrides <: AbstractIncreasingStrides end
+struct StridedLayout <: AbstractStridedLayout end
```

Comment:
This still seems far too fine-grained.  Yes, these are really cool and it's awesome that you can track them through some of our wrapper arrays, but you barely use them for the purposes of dispatching `mul` to BLAS.  Really, BLAS only cares about a handful of things beyond the basic strided-ness:

* Are the columns contiguous? (`stride(A::AbstractMatrix, 1) == 1`)
* Are the rows contiguous? (`stride(A::AbstractMatrix, 2) == 1`)

For the wrapper arrays (reshape and view) to be able to preserve stridedness through their transformations we additionally need to know if the arrays are wholly contiguous (`stride(vec(A), 1) == 1` or `stride(vec(permutedims(A, reverse(ntuple(identity, ndims(A)))), 1) == 1`).

You're going to hate me, but I really think this should just be:

```julia
abstract type AbstractStridedLayout <: MemoryLayout end
struct StridedLayout <: AbstractStridedLayout end
struct ColumnMajorContiguousStridedLayout <: AbstractStridedLayout end
struct RowMajorContiguousStridedLayout <: AbstractStridedLayout end
```

Note that [`gemm_wrapper!`](https://github.com/JuliaLang/julia/blob/029677ab475eecc1f7d1d51d2ae3c1802e36096f/stdlib/LinearAlgebra/src/matmul.jl#L392) _already_ does value-based checking on `stride(A, 1) == 1`, etc.  I really don't see value in having these things encoded into a traitsystem.  To the contrary: I think the additional complexity is just asking for trouble.

Reply:
Let's handle renaming separately, as that's a trivial find/replace change.

So back to what it was originally? Weren't you the one to suggest adding `Increasing/DecreasingStrides`? 

Removing `ColumnMajor` loses functionality: `V = view(A,1:5,1,2,3)` is known to be `DenseColumnMajor` if `A` is `ColumnMajor`, so we can do, for example, `reshape(V, (2,3))` and still use BLAS.

Reply:
I regard relying on overloading of `view` as a no go: the initial motivation for this PR (which came up in other contexts subsequently) was that if `A` is banded then `view(A, ::UnitRange, ::UnitRange)` is also banded.

Now one could override `view` to return a `SubBandedMatrix`, which is what I did initially, but it results in tons of copy-and-paste of subarray.jl. The "correct" solution was to instead link into the current `SubArray` infrastructure. 

I think we should set aside the discussion of simplifying the type hierarchy until others weigh in, as there are a number of different needs here.